File: languages/scala/src/test/scala/fp/scala/ClosuresTest.scala
package fp.scala

import org.junit.Test
import org.junit.Assert._
import scala.collection.mutable.ListBuffer

class ClosuresTest {
  def createAdder(): Int => Int = {
    val x = 3

    // We return an anonymous function that takes one variable, y.
    // We say y is 'bound', because it will be provided when the function is called.
    // x is unbound, or 'free', but since Groovy's anonymous functions are closures,
    // it will get bound to 3 because there is an x in the lexical scope in which the
    // anonymous function was defined.
    (y:Int) => {y + x}
  }

  @Test
  def closuresCanReferenceVariablesInTheirDefiningScope() {
    // x is not in scope here, but the closure retains its binding.
    val adder = createAdder()
    assertEquals(8, adder(5))
  }

  @Test
  def closuresCanIntroduceSideEffects () {
    var myList = ListBuffer("a", "b", "c")
    assertEquals(ListBuffer("b", "c"), removeFirst(() => {myList}))
    assertEquals(ListBuffer("b","c"), myList)                                // Oops...it got modified!
  }

  def removeFirst(computeList: () => ListBuffer[String]) : ListBuffer[String] = {
    val theList = computeList()
    theList.remove(0)
    theList
  }
}

File: languages/java/src/main/java/fp/java/ClosuresTest.java
package fp.java;

import org.junit.Test;
import static org.junit.Assert.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;

public class ClosuresTest {
    Function<Integer, Integer> createAdder() {
        int x = 3;

        // Return a function that adds a number to 3
        return (Integer y) -> {
            return x + y;
        };
    }

    @Test
    public void closuresCanReferenceVariablesInTheirDefiningScope() {
        // x is not in scope here, but the closure retains its binding.
        Function<Integer, Integer> adder = createAdder();
        assertEquals((Integer) 8, adder.apply(5));
    }

    @Test
    public void closuresCanIntroduceSideEffects() {
        // We use ArrayList because it must be mutable
        ArrayList<String> myList = new ArrayList<>(Arrays.asList("a", "b", "c"));

        // We pass in a lambda/closure, which is really an implementation
        // of ListReturner, which takes no args and returns a list.
        assertEquals(Arrays.asList("b", "c"), removeFirst(() -> myList));
        assertEquals(Arrays.asList("b", "c"), myList);                                // Oops...it got modified!
    }

    List<String> removeFirst(ListReturner lr){
        List<String> theList = lr.returnList();
        theList.remove(0);
        return theList;
    }

    /*
        We need a functional interface because our lambda takes no arguments, so
        'Function' does not work.
     */

    @FunctionalInterface
    interface ListReturner {
        List<String> returnList();
    }
}
